pkg/controllers/subnet/subnet_controller.go

```diff
@@ -0,0 +1,225 @@
+type SubnetReconciler struct {
+	Client  client.Client
+	Scheme  *apimachineryruntime.Scheme
+	Service *subnet.SubnetService
+}
+
+func (r *SubnetReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	obj := &v1alpha1.Subnet{}
+	log.Info("reconciling subnet CR", "subnet", req.NamespacedName)
+	metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerSyncTotal, MetricResTypeSubnet)
+
+	if err := r.Client.Get(ctx, req.NamespacedName, obj); err != nil {
+		log.Error(err, "unable to fetch Subnet CR", "req", req.NamespacedName)
+		return ResultNormal, client.IgnoreNotFound(err)
+	}
+
+	if obj.ObjectMeta.DeletionTimestamp.IsZero() {
+		metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerUpdateTotal, MetricResTypeSubnet)
+		if !controllerutil.ContainsFinalizer(obj, servicecommon.FinalizerName) {
+			controllerutil.AddFinalizer(obj, servicecommon.FinalizerName)
+			if err := r.Client.Update(ctx, obj); err != nil {
+				log.Error(err, "add finalizer", "subnet", req.NamespacedName)
+				updateFail(r, &ctx, obj, &err)
+				return ResultRequeue, err
+			}
+			log.V(1).Info("added finalizer on subnet CR", "subnet", req.NamespacedName)
+		}
+		if err := r.Service.CreateOrUpdateSubnet(obj); err != nil {
+			log.Error(err, "operate failed, would retry exponentially", "subnet", req.NamespacedName)
+			updateFail(r, &ctx, obj, &err)
+			return ResultRequeue, err
+		}
+		if err := r.updateSubnetStatus(obj); err != nil {
+			log.Error(err, "update subnet status failed, would retry exponentially", "subnet", req.NamespacedName)
+			updateFail(r, &ctx, obj, &err)
+			return ResultRequeue, err
+		}
+		updateSuccess(r, &ctx, obj)
+	} else {
+		if controllerutil.ContainsFinalizer(obj, servicecommon.FinalizerName) {
+			metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerDeleteTotal, MetricResTypeSubnet)
+			if err := r.Service.DeleteSubnet(obj.UID); err != nil {
+				log.Error(err, "deletion failed, would retry exponentially", "subnet", req.NamespacedName)
+				deleteFail(r, &ctx, obj, &err)
+				return ResultRequeue, err
+			}
+			controllerutil.RemoveFinalizer(obj, servicecommon.FinalizerName)
+			if err := r.Client.Update(ctx, obj); err != nil {
+				log.Error(err, "deletion failed, would retry exponentially", "subnet", req.NamespacedName)
+				deleteFail(r, &ctx, obj, &err)
+				return ResultRequeue, err
+			}
+			log.V(1).Info("removed finalizer", "subnet", req.NamespacedName)
+			deleteSuccess(r, &ctx, obj)
+		} else {
+			log.Info("finalizers cannot be recognized", "subnet", req.NamespacedName)
+		}
+	}
+	return ctrl.Result{}, nil
+}
+
+func (r *SubnetReconciler) updateSubnetStatus(obj *v1alpha1.Subnet) error {
+	nsxSubnets := r.Service.SubnetStore.GetByIndex(servicecommon.TagScopeSubnetCRUID, string(obj.UID))
+	if len(nsxSubnets) == 0 {
+		return errors.New("failed to get subnet from store")
+	}
+	obj.Status.IPAddresses = make([]string, len(nsxSubnets[0].IpAddresses))
+	for i, ip := range nsxSubnets[0].IpAddresses {
```

Comment:
for subnet auto allocated from VPC IPBlock, the NSX Subnet does not include ip_addressed field,
how do we ensure there is value for nsxSubnet.IpAddresses?

Reply:
Indeed, there is no ip_addresses for subnet auto allocated from VPC IPBlock. For this case, how about retrieving subnet status by api `orgs/{{org_id}}/projects/{{pro_id}}/vpcs/{{vpc_id}}/subnets/{{subnet_id}}/status`? Using field `network_address` and a sample result will be like this.

```
        {
            "static_ip_pool_path": "xxx/xxx",
            "static_ip_pool_range": "10.0.0.75-10.0.0.78",
            "gateway_address": "10.0.0.65/28",
            "network_address": "10.0.0.64/28",
            "dhcp_server_address": "10.0.0.66/28",
            "dhcp_ranges": "10.0.0.67-10.0.0.74",
            "ip_address_type": "IPV4"
        }

```

Reply:
Yea, we need to retrieve subnet CIDR and update it under status.

Reply:
Thanks, code has been updated to retrieve subnet CIDR from api `../subnet/status` for scenario that IP addresses are auto allocated from VPC IPBlock.